src/app/vm/vm-creation/vm-deployment.service.ts

```diff
@@ -0,0 +1,228 @@
+import { Injectable } from '@angular/core';
+import { VmCreationState } from './data/vm-creation-state';
+import { Observable } from 'rxjs/Observable';
+import { Subject } from 'rxjs/Subject';
+import { VmService } from '../shared/vm.service';
+import { VirtualMachine, VmStates } from '../shared/vm.model';
+import { AffinityGroupService, InstanceGroupService } from '../../shared/services';
+import { AffinityGroup, AffinityGroupTypes } from '../../shared/models';
+import { SecurityGroup } from '../../security-group/sg.model';
+import { GROUP_POSTFIX, SecurityGroupService } from '../../shared/services/security-group.service';
+import { Utils } from '../../shared/services/utils.service';
+
+
+export type VmDeploymentStage =
+  'STARTED' |
+  'IN_PROGRESS' |
+  'AG_GROUP_CREATION' |
+  'AG_GROUP_CREATION_FINISHED' |
+  'SG_GROUP_CREATION' |
+  'SG_GROUP_CREATION_FINISHED' |
+  'VM_DEPLOYED' |
+  'FINISHED' |
+  'ERROR' |
+  'TEMP_VM';
+
+export const VmDeploymentStages = {
+  STARTED: 'STARTED' as VmDeploymentStage,
+  IN_PROGRESS: 'IN_PROGRESS' as VmDeploymentStage,
+  AG_GROUP_CREATION: 'AG_GROUP_CREATION' as VmDeploymentStage,
+  AG_GROUP_CREATION_FINISHED: 'AG_GROUP_CREATION_FINISHED' as VmDeploymentStage,
+  SG_GROUP_CREATION: 'SG_GROUP_CREATION' as VmDeploymentStage,
+  SG_GROUP_CREATION_FINISHED: 'SG_GROUP_CREATION_FINISHED' as VmDeploymentStage,
+  VM_DEPLOYED: 'VM_DEPLOYED' as VmDeploymentStage,
+  FINISHED: 'FINISHED' as VmDeploymentStage,
+  ERROR: 'ERROR' as VmDeploymentStage,
+  TEMP_VM: 'TEMP_VM' as VmDeploymentStage
+};
+
+export interface VmDeploymentMessage {
+  stage: VmDeploymentStage;
+  [key: string]: any;
+}
+
+export interface VmDeployObservables {
+  deployStatusObservable: Observable<VmDeploymentMessage>;
+  deployObservable: Observable<any>;
+}
+
+@Injectable()
+export class VmDeploymentService {
+  constructor(
+    private affinityGroupService: AffinityGroupService,
+    private instanceGroupService: InstanceGroupService,
+    private securityGroupObservable: SecurityGroupService,
+    private vmService: VmService
+  ) {}
+
+  public deploy(state: VmCreationState): VmDeployObservables {
+    const deployStatusObservable = new Subject<VmDeploymentMessage>();
+
+    return {
+      deployStatusObservable,
+      deployObservable: this.deployObservable(deployStatusObservable, state)
+    }
+  }
+
+  private deployObservable(deployObservable, state): Observable<any> {
+    let deployedVm;
+    let tempVm;
+    return Observable.of(null)
+      .do(() => {
+        deployObservable.next({
+          stage: VmDeploymentStages.STARTED
+        });
```

Comment:
maybe it's better to be able to pass just stage (not as key of an object) to the next if there is no other data?